File: Design-Patterns-Using-JavaScript/AbstractFactory/AbstractFactoryDemo2.js
// Abstract ProductA
class ProductA {
    operationA() {
        console.log("ProductA operationA");
    }
}

// Concrete ProductA1
class ProductA1 extends ProductA {
    operationA() {
        console.log("ProductA1 operationA");
    }

}

// Concrete ProductA2
class ProductA2 extends ProductA {
    operationA() {
        console.log("ProductA2 operationA");
    }
}

// Abstract ProductB
class ProductB {
    operationB() {
        console.log("ProductB operationB");
    }
}

// Concrete ProductB1
class ProductB1 extends ProductB {

    operationB() {
        console.log("ProductB1 operationB");
    }
}

// Concrete ProductB2
class ProductB2 extends ProductB {
    operationB() {
        console.log("ProductB2 operationB");
    }
}

// Abstract Factory
class AbstractFactory {
    createProductA() {
        throw new Error("Abstract method: createProductA");
    }

    createProductB() {
        throw new Error("Abstract method: createProductB");
    }
}

// Concrete Factory1
class ConcreteFactory1 extends AbstractFactory {
    createProductA() {
        return new ProductA1();
    }

    createProductB() {
        return new ProductB1();
    }
}

// Concrete Factory2
class ConcreteFactory2 extends AbstractFactory {
    createProductA() {
        return new ProductA2();
    }

    createProductB() {
        return new ProductB2();
    }
}

// Client code
const factory1 = new ConcreteFactory1();
const productA1 = factory1.createProductA();
const productB1 = factory1.createProductB();
productA1.operationA();
productB1.operationB();

const factory2 = new ConcreteFactory2();
const productA2 = factory2.createProductA();
const productB2 = factory2.createProductB();
productA2.operationA();
productB2.operationB();

/*
ProductA1 operationA
ProductB operationB
ProductA1 operationA
ProductB operationB
*/

File: Design-Patterns-Using-Kotlin/AbstractFactory/AbstractFactoryDemo2.kt

// ProductA interface
interface ProductA {
    fun operationA()
}

// Concrete ProductA1
class ProductA1 : ProductA {
    override fun operationA() {
        println("ProductA1 operationA")
    }
}

// Concrete ProductA2
class ProductA2 : ProductA {
    override fun operationA() {
        println("ProductA2 operationA")
    }
}

// Abstract ProductB
interface ProductB {
    fun operationB()
}

// Concrete ProductB1
class ProductB1 : ProductB {
    override fun operationB() {
        println("ProductB1 operationB")
    }
}

// Concrete ProductB2
class ProductB2 : ProductB {
    override fun operationB() {
        println("ProductB2 operationB")
    }
}

// Abstract Factory
interface AbstractFactory {
    fun createProductA(): ProductA
    fun createProductB(): ProductB
}

// Concrete Factory1
class ConcreteFactory1 : AbstractFactory {
    override fun createProductA(): ProductA {
        return ProductA1()
    }

    override fun createProductB(): ProductB {
        return ProductB1()
    }
}

// Concrete Factory2
class ConcreteFactory2 : AbstractFactory {
    override fun createProductA(): ProductA {
        return ProductA2()
    }

    override fun createProductB(): ProductB {
        return ProductB2()
    }
}

// Client code
fun main() {
    val factory1: AbstractFactory = ConcreteFactory1()
    val productA1: ProductA = factory1.createProductA()
    val productB1: ProductB = factory1.createProductB()
    productA1.operationA()
    productB1.operationB()

    val factory2: AbstractFactory = ConcreteFactory2()
    val productA2: ProductA = factory2.createProductA()
    val productB2: ProductB = factory2.createProductB()
    productA2.operationA()
    productB2.operationB()
}

/*
ProductA1 operationA
ProductB1 operationB
ProductA2 operationA
ProductB2 operationB
*/